File: September/26/combos.py
# LeetCode 77: Combinations
# Explanation:
# 1. Use backtracking to build combinations.
# 2. At each step, choose a number or skip it.
# 3. Stop when we have k numbers.
# Time Complexity: O(C(n, k) * k) because each combination requires building.
# Space Complexity: O(k) recursion depth.

class Solution:
    def combine(self, n: int, k: int) -> list[list[int]]:
        res = []
        
        def backtrack(start, comb):
            if len(comb) == k:
                res.append(comb[:])
                return
            for i in range(start, n + 1):
                comb.append(i)
                backtrack(i + 1, comb)
                comb.pop()
        
        backtrack(1, [])
        return res

    def combinePrune(self, n: int, k: int) -> list[list[int]]:
        res = []
        
        def backtrack(start, comb):
            if len(comb) == k:
                res.append(comb[:])
                return
            for i in range(start, n - (k - len(comb)) + 2):
                comb.append(i)
                backtrack(i + 1, comb)
                comb.pop()
        
        backtrack(1, [])
        return res
    
    combineBuiltIn=lambda s,n,k:__import__('itertools').combinations(range(1,n+1),k)

# Example usage:
# sol = Solution()
# print(sol.combine(4, 2))  # [[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]

File: 2025/September/26/combos.cpp
// LeetCode 77: Combinations
// Explanation:
// 1. Use backtracking to generate all valid combinations.
// 2. Stop when current combination has size k.
// Time Complexity: O(C(n, k) * k)
// Space Complexity: O(k)

class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> comb;
        backtrack(1, n, k, comb, res);
        return res;
    }

    void backtrack(int start, int n, int k, vector<int>& comb, vector<vector<int>>& res) {
        if (comb.size() == k) {
            res.push_back(comb);
            return;
        }
        for (int i = start; i <= n; i++) {
            comb.push_back(i);
            backtrack(i + 1, n, k, comb, res);
            comb.pop_back();
        }
    }
};

// Example usage:
// Solution sol;
// auto ans = sol.combine(4, 2);
// for (auto& v : ans) {
//     for (int x : v) cout << x << " ";
//     cout << endl;
// }
// // [[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]
